### PR TITLE
[SPARK] fix removePathPattern behaviour

### PR DESCRIPTION
**Pull request by pawel-big-lebowski**
### Problem

removepath pattern feature is not applied all the time. The method is called when constructing `DatasetIdentifier` through `PathUtils` which is not the case all the time. 

Closes: #2335

### Solution
 
* Move removePattern to another place in codebase which is always run,
* Write integration test for this. 

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] Your comment includes a one-liner for the changelog about the specific purpose of the change (_if necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [ ] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/DatasetIdentifier.java [hunk 1]
```diff
@@ -1,44 +0,0 @@
...
-*/
-
-package io.openlineage.spark.agent.util;
-
-import java.util.LinkedList;
-import java.util.List;
-import lombok.Value;
-
-@Value
-public class DatasetIdentifier {
```
Comment (mobuchowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/DatasetIdentifier.java:13]: Duplicate?

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/RemovePathPatternUtils.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ * Utility class to handle removing path patterns in dataset names. Given a configured regex pattern
+ * with "remove" group defined, class methods run regex replacements on all the datasets available
+ * within the event
+ */
+public class RemovePathPatternUtils {
+  public static final String REMOVE_PATTERN_GROUP = "remove";
+  public static final String SPARK_OPENLINEAGE_DATASET_REMOVE_PATH_PATTERN =
+      "spark.openlineage.dataset.removePath.pattern";
+
+  public static List<OutputDataset> removeOutputsPathPattern(
```
Comment (mobuchowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/RemovePathPatternUtils.java:27]: Maybe `List<? extends Dataset>` and implement this generically with passed dataset builder? Nit, do what you think is best.

**Review by mobuchowski — APPROVED** [review]
One comment but lgtm.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/DatasetIdentifier.java (hunk 1 above)
Comment (pawel-big-lebowski) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/DatasetIdentifier.java:13]: this got moved to java client some time ago

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/RemovePathPatternUtils.java (hunk 2 above)
Comment (pawel-big-lebowski) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/util/RemovePathPatternUtils.java:27]: I have tried this approach but did not succeed. 
Creating `OutputDataset` and `InputDataset` differs significantly and these have to be two separate methods. 
The iterating part, that goes through the list, is common.
So a common generic method would need to call methods  per `OutputDataset` or `InputDataset`, which didn't look as an improvement to me.

**Pull request merged**